### PR TITLE
[#174]added warning screen + StorageChecker.kt

### PR DESCRIPTION
**Pull request by gasstan**
This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Manual tests: Did you update the [manual tests](../blob/main/docs/testing/manual_testing) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed._ Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules.
- [ ] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes?
- [x] Screenshots: Did you provide before and after UI screenshots in the description of this pull request?  _This is only applicable for changes that modify the UI._
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [x] Manual tests: Did you review the manual tests?
- [ ] How is code coverage affected? _We encourage you to compare coverage before and after changes and when possible, leaving it in a better place._
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [x] Run the app: Did you run the app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Comment by gasstan** [comment]
still waiting for ui/ux specs

**Review by ccjernigan — COMMENTED** [review]
This is a good start!

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/home/viewmodel/WalletViewModel.kt [hunk 1]
```diff
@@ -90,6 +91,13 @@ class WalletViewModel(application: Application) : AndroidViewModel(application)
         emitAll(StandardPreferenceKeys.IS_USER_BACKUP_COMPLETE.observe(preferenceProvider))
     }
 
+    val isEnoughSpace = flow { emit(StorageChecker.isEnoughSpace()) }
```
[the remark this answers is not in the document]
Comment (gasstan) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/home/viewmodel/WalletViewModel.kt:94]: i've created new ViewModel to manage storage space data

File: ui-lib/src/main/res/ui/warning/values/strings.xml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="not_enough_space_title">Not enough space!</string>
+    <string name="not_enough_space_description">You need approximately %1$s gig of space while synchronizing the Zcash blockchain, but only 300 megs once done. Syncing will stay paused until more space is available.</string>
```
Comment (ccjernigan) [inline ui-lib/src/main/res/ui/warning/values/strings.xml:4]: Can you update the strings to use xliff?  Some of our other strings have examples.  It'll help once we need to hire localization help.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowKtx.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.WhileSubscribed
+import kotlinx.coroutines.flow.stateIn
+
+/**
+ * Converts Flow to StateFlow with default sharing strategy and initial value set to null
+ */
+fun <T> Flow<T>.toStateFlow(
```
Comment (ccjernigan) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowKtx.kt:14]: I'd prefer we not introduce this as a one-off.  We either need to use this everywhere (thus modifying all our view models) or nowhere.

I think there's value in keeping our existing pattern of being explicit with:

```
.stateIn(
        viewModelScope,
        SharingStarted.WhileSubscribed(ANDROID_STATE_FLOW_TIMEOUT),
        null
    )
```

The reason I'm not 100% convinced this syntactic sugar is the right move is that some places we specifically use different behaviors like `WalletCoordinator` and `SupportViewModel`.  I think being explicit makes it obvious that someone really thought through the usage of the StateFlow.

Thoughts?

**Review by ccjernigan — COMMENTED** [review]
Thanks for making this enhancement!  I've made a few inline comments.

Can you include a screenshot as a GitHub comment?  That helps the team, e.g.  geffenz, also review the design aspects.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowKtx.kt (hunk 3 above)
Comment (gasstan) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowKtx.kt:14]: Yes i agree with what you saying, but i see this extension just as a nice way to remove unwanted boilerplate.  On the other hand i think you're right that being explicit makes it more obvious.

**Comment by gasstan** [comment]
(not pixel perfect yet)
Screenshot (Oct 3, 2022 11:00:07)

**Review by ccjernigan — APPROVED** [review]

**Pull request merged**